cmd/runtimetest/main.go

```diff
@@ -609,21 +609,71 @@ func validateMountsExist(spec *rspec.Spec) error {
 		mountsMap[mountInfo.Mountpoint] = append(mountsMap[mountInfo.Mountpoint], m)
 	}
 
-	for _, specMount := range spec.Mounts {
-		if specMount.Type == "bind" || specMount.Type == "rbind" {
+	for _, configMount := range spec.Mounts {
+		if configMount.Type == "bind" || configMount.Type == "rbind" {
 			// TODO: add bind or rbind check.
 			continue
 		}
 
 		found := false
-		for _, sysMount := range mountsMap[filepath.Clean(specMount.Destination)] {
-			if err := mountMatch(specMount, sysMount); err == nil {
+		for _, sysMount := range mountsMap[filepath.Clean(configMount.Destination)] {
+			if err := mountMatch(configMount, sysMount); err == nil {
 				found = true
 				break
 			}
 		}
 		if !found {
-			return fmt.Errorf("Expected mount %v does not exist", specMount)
+			return fmt.Errorf("Expected mount %v does not exist", configMount)
+		}
+	}
+
+	return nil
+}
+
+func validateMountsOrder(spec *rspec.Spec) error {
+	if runtime.GOOS == "windows" {
+		logrus.Warnf("mounts order validation not yet implemented for OS %q", runtime.GOOS)
+		return nil
+	}
+
+	mountInfos, err := mount.GetMounts()
+	if err != nil {
+		return err
+	}
+
+	type mountOrder struct {
+		Order  int
+		Root   string
+		Dest   string
+		Source string
+	}
+	mountsMap := make(map[string][]mountOrder)
+	for i, mountInfo := range mountInfos {
+		m := mountOrder{
+			Order:  i,
+			Root:   mountInfo.Root,
+			Dest:   mountInfo.Mountpoint,
+			Source: mountInfo.Source,
+		}
+		mountsMap[mountInfo.Mountpoint] = append(mountsMap[mountInfo.Mountpoint], m)
+	}
+	current := -1
+	for _, configMount := range spec.Mounts {
+		mounts := mountsMap[configMount.Destination]
+		for _, mount := range mounts {
+			source := mount.Source
+			for _, option := range configMount.Options {
+				if option == "bind" || option == "rbind" {
+					source = mount.Root
+				}
+			}
+			if source == configMount.Source {
+				if current > mount.Order {
+					return fmt.Errorf("%s is not mounted in order", configMount.Source)
```

Comment:
You could return a more informative message here if `current` was a`*mountOrder`.